omnipaxos_storage/src/persistent_storage.rs

```diff
@@ -0,0 +1,587 @@
+            "Cannot create new instance, database already exists in {}",
+            path
+        ));
+
+        Self::open(storage_config)
+    }
+}
+
+impl<T, S> Storage<T, S> for PersistentStorage<T, S>
+where
+    T: Entry + Serialize + for<'a> Deserialize<'a>,
+    S: Snapshot<T> + Serialize + for<'a> Deserialize<'a>,
+{
+    fn append_entry(&mut self, entry: T) -> u64 {
+        let entry_bytes = bincode::serialize(&entry).expect("Failed to serialize log entry");
+        let offset = self
+            .commitlog
+            .append_msg(entry_bytes)
+            .expect("Failed to append log entry");
+        self.commitlog.flush().expect("Failed to flush Commitlog"); // ensure durable writes
+        offset + 1 // +1 as commitlog returns the offset the entry was appended at, while we should return the index that the entry got in the log.
+    }
+
+    fn append_entries(&mut self, entries: Vec<T>) -> u64 {
+        let serialized = entries
+            .into_iter()
+            .map(|entry| bincode::serialize(&entry).expect("Failed to serialize log entries"));
+        let offset = self
+            .commitlog
+            .append(&mut MessageBuf::from_iter(serialized))
+            .expect("Falied to append log entries");
+        self.commitlog.flush().expect("Failed to flush Commitlog"); // ensure durable writes
+        offset.first() + offset.len() as u64
+    }
+
+    fn append_on_prefix(&mut self, from_idx: u64, entries: Vec<T>) -> u64 {
+        self.commitlog
+            .truncate(from_idx)
+            .expect("Failed to truncate log");
+        self.append_entries(entries)
+    }
+
+    fn get_entries(&self, from: u64, to: u64) -> Vec<T> {
+        let buffer = self
+            .commitlog
+            .read(from, ReadLimit::default())
+            .expect("Failed to read from replicated log");
+        let mut entries = Vec::<T>::with_capacity((to - from) as usize);
+        let mut iter = buffer.iter();
+        for _ in from..to {
+            match iter.next() {
+                Some(msg) => {
+                    entries.push(
+                        bincode::deserialize(msg.payload())
+                            .expect("Failed to deserialize log entries"),
+                    );
+                }
+                None => {
+                    return vec![]; // early return here
+                }
+            }
+        }
+        entries
+    }
+
+    fn get_log_len(&self) -> u64 {
+        self.commitlog.next_offset()
+    }
+
+    fn get_suffix(&self, from: u64) -> Vec<T> {
+        self.get_entries(from, self.commitlog.next_offset())
+    }
+
+    fn get_promise(&self) -> Ballot {
+        #[cfg(feature = "rocksdb")]
+        {
+            let promised = self.rocksdb.get(NPROM).expect("Failed to retrieve 'NPROM'");
+            match promised {
+                Some(prom_bytes) => {
+                    let b_store: BallotStorage = FromBytes::read_from(prom_bytes.as_slice())
```

Comment:
It is also possible to use the zerocopy-derived struct directly as they are implemented for your types.
```rust
let b_store = BallotStorage::read_from(prom_bytes.as_slice())).expect(....);
```